BUG: Fix crash on 0d return value in apply_along_axis

eric-wieser opened the pull request:
Also:
ENH: Support arbitrary dimensionality of return value
MAINT: remove special casing

Now supports
```python
>>> data = np.arange(6).reshape(2, 3)
>>> data
array([[0, 1, 2],
       [3, 4, 5]])
>>> np.apply_along_axis(np.diag, -1, data)  #would previously crash
array([[[0, 0, 0],
        [0, 1, 0],
        [0, 0, 2]],

       [[3, 0, 0],
        [0, 4, 0],
        [0, 0, 5]]])
>>> np.apply_along_axis(lambda x: np.array(np.sum(x)), 0, data)  #would previously crash
array([3, 5, 7])
```

Addresses [discussion from #8363](https://github.com/numpy/numpy/pull/8363#issuecomment-270027102)

eric-wieser committed:
BUG: Fix crash on 0d return value in apply_along_axis

Also:
ENH: Support arbitrary dimensionality of return value
MAINT: remove special casing

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -74,65 +75,53 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    # save the first result
+    outarr_view[tuple(ind)] = res
+    k = 1
+    while k < Ntot:
```
Is this really likely to be faster than using `ndindex` here?

eric-wieser committed:
MAINT: Use np.ndindex, which seems just as efficient

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -74,65 +76,42 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    inarr_view = transpose(arr, dims_in[:axis] + dims_in[axis+1:] + [axis])
+
+    # indices
+    inds = ndindex(inarr_view.shape[:nd-1])
```
To be honest, most of this function could probably be replaced with a carefully constructed `nditer` that does the transposes for us, but this is probably good enough for now

eric-wieser committed:
BUG: Call __array_prepare__ before __array_wrap__

eric-wieser commented:
Right now, this happens:
```
# silent broadcasting
>>> apply_along_axis(lambda x: 1 if x.all() else [2, 3], 0, np.array([[False, True, True]]))))
array([[2, 1, 1],
       [3, 1, 1]])
# same function, different order
>>> apply_along_axis(lambda x: 1 if x.all() else [2, 3], 0, np.array([[True, False, True]]))))
ValueError: setting an array element with a sequence.
```

Would it be better if both raised `ValueError: shape must be the same on each invocation`, at the expense of a check on every iteration?

eric-wieser commented:
#5875 is rearing its head in the errors here...

eric-wieser commented:
This fails when `f()` returns a `matrix`, because `matrix`s violate the assumption that `arr[0].shape == arr.shape[1:]`

eric-wieser committed:
BUG: Work around evil matrix.__array_prepare__

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -100,7 +101,11 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    if not isinstance(res, matrix):
+        outarr = res.__array_prepare__(outarr)
+    if outshape != outarr.shape:
+        raise ValueError('__array_prepare__ should not change the shape of the resultant array')
```
Is this assumption supposed to be valid? How do we deal with `matrix` violating it elsewhere?

rgommers commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -28,13 +28,23 @@ def test_3d(self):
+        )
+
+        result = apply_along_axis(double, 1, m)
+        assert isinstance(result, np.matrix)
```
Note that assert shouldn't be used at all, use np.testing.assert_ instead (applies to multiple lines below as well)

eric-wieser replied:
In my defense, those other lines were not part of this patch. Do you want me to fix all cases of assert in this file anyway?

eric-wieser replied:
Or should this be `self.assertIsInstance`?

eric-wieser replied:
Changed to `assert_(isinstance(...))`, because that's what other tests do. Switching to `self.assertIsInstance` can be discussed elsewhere

rgommers replied:
> In my defense, those other lines were not part of this patch. Do you want me to fix all cases of assert in this file anyway?

if you could, that'd be useful

rgommers replied:
``self.assertIsInstance`` seems fine to use if you prefer that, it has a sane implementation in both ``unittest`` and ``unittest2``.

eric-wieser replied:
@rgommers: I've patched the other ones in that file

eric-wieser commented:
Ugh, this needs a patch in [`ma.apply_along_axis`](https://github.com/numpy/numpy/blob/v1.12.0/numpy/ma/extras.py#L377-L461) as well.

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    outarr = zeros(outshape, res.dtype)
+    # matrices call reshape in __array_prepare__, and this is apparently ok!
+    if not isinstance(res, matrix):
+        outarr = res.__array_prepare__(outarr)
```
Can someone comment on whether this is actually correct, or if it can just be omitted altogether

`__array_prepare__` seems to always do something screwy for builtin array subclasses. It's not clear to me what the contract entails, and whether I'm violating it, or `np.matrix` and `np.ma.MaskedArray` are

shoyer replied:
@mhvk may know

eric-wieser replied:
I think `MaskedConstant.__array_prepare__` was broken - fixed in #8508. I think a corollary of this is that `np.ma.apply_along_axis` can basically defer now.

shoyer commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    # save the first result
+    outarr_view[ind0] = res
+    for ind in inds:
+        outarr_view[ind] = asanyarray(func1d(inarr_view[ind], *args, **kwargs))
```
I recognize that this works to update the view inplace, but it still makes me a little nervous. It would be more direct and obviously correct to do the transposing afterwards.

This would change the memory layout of the result array. Transposing afterwards would ensure that array is written with contiguous writes, which could result in a marginal performance improvement (at least for this function).

eric-wieser replied:
Yep, my goal here was to leave the memory layout of the result array unaffected when compared to the old version.

In particular, the old code would always give a contiguous output, something that may have been relied upon

shoyer replied:
I think we reserve the right to change the memory layout of arrays returned by NumPy functions unless it is documented as part of the API. Users who truly rely on contiguous output should be using `np.ascontiguousarray`.

eric-wieser replied:
I guess transposing afterwards would fix the above issue

seberg replied:
We have done it before (e.g. even indexing I did it). If it is obvious/possible, then "the same as before" should be preferred most of the time I would say.

shoyer replied:
I agree that keeping the same memory layout should be preferred in general. But in this case, keeping the same memory layout requires jumping through some hoops, and the resulting code is much less obvious. So I would lean towards changing it.

seberg replied:
Sure, I did not even check what is being checked here, just saying what my take on the policy is :).

eric-wieser replied:
I'd argue that the correctness is more obvious this way around, because otherwise the inverse transposed order is needed, which is more troublesome to generate

shoyer replied:
I mostly like transposing afterwards more. I don't care so much about the memory layout.

eric-wieser replied:
Actually I think you're right, it is simpler - the transpose order becomes `dims_out[:axis] + dims_out[-res.ndim:] + dims_out[axis+1:]`

eric-wieser replied:
Should the transpose happen before or after the `__array_wrap__`?

eric-wieser replied:
Ok, I'm sold on your suggestion, but for a different reason: when `res.__array_prepare__` is called, the axes of `res` (a single output) and the final output are paired as if they were broadcasted, which seems desirable for any kind of axis metadata copying

shoyer reviewed (commented):
I really like the approach here, thanks for tackling this clean-up!

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+
+    # arr, with the iteration axis at the end
+    dims_in = list(range(nd))
+    inarr_view = transpose(arr, dims_in[:axis] + dims_in[axis+1:] + [axis])
```
Actually, only the output `transpose` does

shoyer replied:
As @seberg writes in that mailing list discussion, transpose on base numpy arrays always returns a view, never a copy.

shoyer replied:
But this might indeed break in surprising ways for ndarray subclasses

eric-wieser replied:
@shoyer: Would an acceptable compromise be to not call `__array_prepare__` if `tranpose` is a copy?

eric-wieser committed:
TST: Verify apply_along_axis now works on masked arrays

Note that this is not a full subsitute for np.ma.apply_along_axis,
as that allows functions to return a mix of np.ma.masked and scalars

eric-wieser committed:
MAINT: Transpose the result, rather than working with a transposed view

.transpose does not specify that it must return a view, so subclasses
(like np.ma.array) could otherwise break this.

This exposes some more need for matrix special casing.

eric-wieser commented:
@shoyer: Transposed moved, which now gives this `masked_array` support, leading into #8511

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     if axis < 0:
         axis += nd
-    if (axis >= nd):
+    if axis >= nd:
```
while we're at it, also check that axis is not still negative?

eric-wieser replied:
I like the idea of killing this altogether, and delegating parsing `axis` to `moveaxis`

eric-wieser replied:
Alarmingly, I can't even find a precedent where numpy checks for this

mhvk replied:
@eric-wieser - when you add the changelog, could you also move this check up to before `axis` is possibly changed and check for too-negative as well? That way, the error message gives the actual input rather than something that's already changed. I.e.,
```
if axis < -nd or axis >= nd:
    raise ...

if axis < 0:
    axis += nd
```

eric-wieser replied:
@mhvk: Done

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+
+    # permutation of axes such that out = pre_arr.transpose(pre_permute)
+    pre_dims = list(range(pre_arr.ndim))
+    pre_permute = pre_dims[:axis] + list(roll(pre_dims[axis:], res.ndim))
```
Maybe a bit faster (no list->ndarray->list) and clearer (to me at least):
```
if res.ndim > 0:
   pre_permute = pre_dims[:axis] + pre_dims[-res.ndim:] + pre_dims[axis+red.ndim:]
```
(you'd need to rename `pre_dims` to `pre_permute` if you want the transpose at the end even for `res.ndim=0`)

eric-wieser replied:
Yeah, that line was inspired by a talk extolling the virtues of `std::rotate`.

Had that before, but without the if - as you correctly spot, it fails when `res.ndim == 0`. Not a huge fan of special casing that.

Either way, I reckon moveaxis will do the job here too.

eric-wieser replied:
Or how about just
```
pre_permute = pre_dims[:axis] + pre_dims[nd-res.ndim:] + pre_dims[axis+red.ndim:]
```

Adding the `nd` there fixes the error when res.ndim= 0

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+        pre_arr = res.__array_wrap__(pre_arr)
+
+        # finally, rotate the inserted axes back to where they belong
+        return transpose(pre_arr, pre_permute)
```
For speed for the more common case (and required with my comment above):
```
return pre_arr if res.ndim == 0 else transpose(pre_arr, pre_permute)
```

mhvk commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
-            outarr = outarr.squeeze(axis)
-        return outarr
+        # matrices have to be transposed first, because they collapse dimensions!
+        out_arr = transpose(pre_arr, pre_permute)
```
same as above

mhvk commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -28,13 +28,23 @@ def test_3d(self):
         result = apply_along_axis(double, 0, m)
-        assert isinstance(result, np.matrix)
+        assert_(isinstance(result, np.matrix))
+        assert_array_equal(
```
Just on one line?
```
assert_array_equal(result, expected)
```

mhvk commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -28,13 +28,23 @@ def test_3d(self):
         assert_array_equal(
-            result, np.matrix([[0, 2], [4, 6]])
+            result, expected
         )
```
same

mhvk commented:
@eric-wieser - this is very nice indeed, I like the generality, and the care for subclasses. My comments are all small, but hopefully make this a little better still.

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
-        return outarr
+
+    # arr, with the iteration axis at the end
+    in_dims = list(range(nd))
```
Oh, good point. `moveaxis` works with multiple axes too, so should work in both cases.

shoyer replied:
For restoring axes, you could maybe write something like:
```python
moveaxis(pre_arr, [i + nd - 1 for i in range(res.ndim)],
         [i + axis for i in range(res.ndim)])
```

I think this is a slight improvement over building up the permutation axes for transpose directly.

eric-wieser replied:
I think `moveaxis` should learn a shorthand for dest such that this would work:
```
moveaxis(pre_arr, [i + nd - 1 for i in range(res.ndim)], axis)
```

eric-wieser replied:
@shoyer: Based on @mhvk's revised opinion, are you happy for this to remain as it is? Any further changes blocking this being merged?

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    # remove the requested axis, and add the new ones on the end.
+    # laid out so that each write is contiguous.
+    # for a tuple index inds, pre_arr[inds] = func1d(inarr_view[inds])
+    pre_shape = arr.shape[:axis] + arr.shape[axis+1:] + res.shape
```
Nice!

eric-wieser commented:
I'll do another round of clean up tomorrow to address those

mhvk commented:
> I'll do another round of clean up tomorrow to address those

I think I'd go for your approach rather than `moveaxis` as it is not all that obviously handier in passing in multiple axes.

eric-wieser commented on numpy/lib/shape_base.py:
```diff
@@ -103,11 +103,10 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     # arr, with the iteration axis at the end
-    in_dims = list(range(nd))
-    inarr_view = transpose(arr, in_dims[:axis] + in_dims[axis+1:] + [axis])
+    inarr_view = moveaxis(arr, axis, -1)
```
To me, this seems like it would be a fair bit slower, and isn't much clearer.

eric-wieser replied:
Ooh, how about `np.r_[:axis, axis+1:nd, axis]`

eric-wieser commented:
Keep changing my mind on this one. I think it's best left without `moveaxis`. It doesn't add much to clarity in the second case, and I think there's a benefit to having both transposes written in the same style. Also, building the axis lists directly should be a lot faster than incurring list comprehensions, then the type conversion and error checking that occurs in `moveaxis`.

`np.r_` added some minor conciseness, but at the expense of some performance

I agree that the `roll` was a little cryptic in the second case.

mhvk commented:
@eric-wieser - sorry for having brought up `moveaxis` when I had also concluded that in the end it was not as good an idea -- it works well for the first case, not so well for the second.

mhvk commented:
I had another look at the code and looks all OK.

eric-wieser commented:
Does this need a release note?

mhvk commented:
Always good to add in the bug-fix section!

shoyer commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -55,9 +61,45 @@ class MinimalSubclass(np.ndarray):
         assert_array_equal(res, np.array([6, 6, 6]).view(MinimalSubclass))
 
+    def test_0d_array(self):
+        class MinimalSubclass(np.ndarray):
```
These tests for shape behavior shouldn't test subclasses at the same time. Can you move the subclass checks (if they're really necessary) to their own tests?

eric-wieser replied:
How about `test_0d_array(self, cls=np.ndarray)`, and having the subclass tests call that with a subclass?

shoyer replied:
Sure, something like that sounds reasonable to me.

eric-wieser replied:
@shoyer: Done

shoyer reviewed (commented):
I have a minor suggestion for cleaning up the test suite. Otherwise, this looks good to me.

eric-wieser committed:
MAINT: Improve error-checking of axis argument

Copied from the implementation in core.shape_base.stack

eric-wieser committed:
DOC: Update 1.13.0 release notes with apply_along_axis changes

eric-wieser commented:
Ok, tests fixed, axis checking added, and release notes done too!

shoyer commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -50,13 +56,50 @@ def minimal_function(array):
+
+    def test_axis_insertion(self, cls=np.ndarray):
+        a = np.arange(18).reshape((6, 3))
+        res = apply_along_axis(lambda x: np.diag(x).view(cls), 0, a)
```
Reading over these tests one last time, it occurs to me that we should be more comprehensive for testing insertion of new axes.

Right now you only test inserting at the start, but that doesn't really exercise the permutation logic in a comprehensive way. Let's also test inserting in the middle and at the end. Also, diag is axes order invariant, so none of the tests verify that new axes are inserted in the right order.

eric-wieser replied:
Maybe outer product with itself would be better than diag then?

shoyer replied:
I think the outer product is still invariant to transposes. Something like `x[:, np.newaxis] - x[np.newaxis, :]` would work, though.

shoyer replied:
On re-reading: yes, outer product with its **reverse**, would be fine

eric-wieser replied:
You didn't miss it the first time, I did indeed suggest a plain outer product beforehand, and realized my mistake

eric-wieser replied:
You didn't miss it the first time, I did indeed suggest a plain outer product beforehand, and realized my mistake. This is now done too.

shoyer commented on numpy/lib/tests/test_shape_base.py:
```diff
@@ -50,13 +56,50 @@ def minimal_function(array):
+        res = apply_along_axis(lambda x: np.diag(x).view(cls), 0, a)
+        assert_(isinstance(res, cls))
+        assert_equal(res.ndim, 3)
+        assert_array_equal(res[:,:,0], np.diag(a[:,0]).view(cls))
```
A cleaner way to check this sort of thing is to construct the desired result array and just call `assert_array_equal` (e.g., `np.stack([np.diag(a[:,i]) for i in range(3)], axis=-1).view(cls)`). Otherwise it's easy to overlook one part of the equality check (e.g., shape, in this case).

eric-wieser replied:
@shoyer: I was hoping I could be done with this, but you are definitely right. I'll fix those things up soon

eric-wieser replied:
Ok, done.

shoyer reviewed (approved):
LGTM. Will merge this shortly if no one else has further comments.

shoyer commented:
@eric-wieser could you please clean up the git history a little bit? Thanks

eric-wieser commented:
@shoyer: Fair, I'll give that a go over the weekend.

eric-wieser commented:
@shoyer: Down from 13 commits to 8 commits. Look ok?

shoyer commented:
I would usually squash down to one commit, but this is fine.
